File: ios/PersonalSite/Core/SiteAPIClient.swift
import Foundation

/// 站点 Web API 的薄封装：GET JSON，以及 POST /api/ask 的 SSE 入口。
/// baseURL 来自 Config（本地开发指向 http://127.0.0.1:3000）。
struct SiteAPIClient: Sendable {
    var baseURL: URL = Config.siteBaseURL
    var session: URLSession = .shared

    enum APIError: Error, Equatable {
        case badStatus(Int)
        /// 限流：带服务端的 Retry-After（秒），可能没有。
        case rateLimited(retryAfterSeconds: Int?)
    }

    /// GET 一个 JSON 接口并解码；jsonb 外层行与 camelCase 内容的解码策略由具体模型承担。
    func get<T: Decodable>(_ path: String, queryItems: [URLQueryItem] = []) async throws -> T {
        let normalizedPath = path.hasPrefix("/") ? String(path.dropFirst()) : path
        var components = URLComponents(url: baseURL.appending(path: normalizedPath), resolvingAgainstBaseURL: false)
        if !queryItems.isEmpty { components?.queryItems = queryItems }
        guard let url = components?.url else { throw URLError(.badURL) }
        let (data, response) = try await session.data(from: url)
        try validate(response)
        return try JSONDecoder().decode(T.self, from: data)
    }

    /// POST /api/ask：返回原始 SSE 事件序列；事件到业务模型的映射见 AskStreamEvent。
    func askEvents(
        conversationId: String,
        visitorId: String,
        question: String,
        scope: AskScope
    ) async throws -> SSEEventSequence<AsyncLineSequence<URLSession.AsyncBytes>> {
        var request = URLRequest(url: baseURL.appending(path: "api/ask"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
        request.httpBody = try JSONEncoder().encode(AskRequestBody(
            conversationId: conversationId,
            visitorId: visitorId,
            question: question,
            scope: scope
        ))
        let (bytes, response) = try await session.bytes(for: request)
        if let http = response as? HTTPURLResponse, http.statusCode == 429 {
            throw APIError.rateLimited(retryAfterSeconds: http.value(forHTTPHeaderField: "Retry-After").flatMap(Int.init))
        }
        try validate(response)
        return SSEStream.events(from: bytes)
    }

    private func validate(_ response: URLResponse) throws {
        guard let http = response as? HTTPURLResponse else { throw URLError(.badServerResponse) }
        guard (200..<300).contains(http.statusCode) else { throw APIError.badStatus(http.statusCode) }
    }
}

File: android/app/src/main/java/com/personalsite/core/SiteApiClient.kt
package com.personalsite.core

import com.personalsite.models.AskRequestBody
import com.personalsite.models.AskScope
import java.io.IOException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.serialization.json.Json
import kotlinx.serialization.encodeToString
import kotlinx.serialization.decodeFromString
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

/**
 * 站点 Web API 的薄封装：GET JSON，以及 POST /api/ask 的 SSE 入口。
 * baseUrl 来自 Config（模拟器本地开发指向 http://10.0.2.2:3000）。
 */
class SiteApiClient(
    private val baseUrl: String = Config.siteBaseUrl,
    private val client: OkHttpClient = sharedClient,
) {
    sealed class ApiError : IOException() {
        data class BadStatus(val status: Int) : ApiError()
        /** 限流：带服务端的 Retry-After（秒），可能没有。 */
        data class RateLimited(val retryAfterSeconds: Int?) : ApiError()
    }

    /** GET 一个 JSON 接口并解码；外层行与 camelCase 内容的解码策略由具体模型承担。 */
    suspend inline fun <reified T> get(path: String, query: Map<String, String> = emptyMap()): T {
        val url = buildUrl(path, query)
        val request = Request.Builder().url(url).get().build()
        val body = execute(request).use { it.body.string() }
        return json.decodeFromString(body)
    }

    /** POST /api/ask：返回 SSE 事件流；事件到业务模型的映射见 AskStreamEvent。 */
    fun askEvents(
        conversationId: String,
        visitorId: String,
        question: String,
        scope: AskScope,
    ): Flow<SSEEvent> = flow {
        val payload = json.encodeToString(
            AskRequestBody(
                conversationId = conversationId,
                visitorId = visitorId,
                question = question,
                scope = scope,
            )
        )
        val request = Request.Builder()
            .url(buildUrl("/api/ask"))
            .post(payload.toRequestBody("application/json".toMediaType()))
            .header("Accept", "text/event-stream")
            .build()
        execute(request).use { response ->
            val source = response.body.source()
            val parser = SSEParser()
            while (true) {
                val line = source.readUtf8Line() ?: break
                parser.process(line)?.let { emit(it) }
            }
            parser.finish()?.let { emit(it) }
        }
    }.flowOn(Dispatchers.IO)

    fun buildUrl(path: String, query: Map<String, String> = emptyMap()): HttpUrl {
        val normalized = path.removePrefix("/")
        val builder = "$baseUrl/$normalized".toHttpUrl().newBuilder()
        query.forEach { (name, value) -> builder.addQueryParameter(name, value) }
        return builder.build()
    }

    @PublishedApi
    internal fun execute(request: Request): okhttp3.Response {
        val response = client.newCall(request).execute()
        if (response.code == 429) {
            val retryAfter = response.header("Retry-After")?.toIntOrNull()
            response.close()
            throw ApiError.RateLimited(retryAfter)
        }
        if (response.code !in 200..299) {
            val status = response.code
            response.close()
            throw ApiError.BadStatus(status)
        }
        return response
    }

    companion object {
        val json = Json { ignoreUnknownKeys = true }

        /** 单例 OkHttpClient：连接池复用；SSE 长连接不设读超时。 */
        val sharedClient: OkHttpClient by lazy {
            OkHttpClient.Builder()
                .readTimeout(0, java.util.concurrent.TimeUnit.MILLISECONDS)
                .build()
        }
    }
}
